### PR TITLE
ED-1269 ch lookup epic fe

### PR DESCRIPTION
**Pull request by stevenburnell-dit**
Note: 
Although this include CSS to support the new Confirm company design it does not include the template changes. They have been held back due to questions over whether the new design should be included in this, or added as part of the Registration flow upgrade project.

### CODE REVIEW

File: enrolment/static/javascripts/index.js [hunk 1]
```diff
@@ -416,20 +423,39 @@ GOVUK.components = (new function() {
...
     this._private.$field = $field || $input; // Allows a different form field to receive value.
+    this._private.$form = $input.parents("form");
+    this._private.$errors = $(".errors", this._private.$form);
+    
+    // Custom error handling.
+    this._private.$form.on("submit.CompaniesHouseNameLookup", function(e) {
+      // If no input or no company selected
+      if(instance._private.$field.val() === "") {
+        e.preventDefault();
+        instance._private.$errors.append("<p>Check that you entered the company name correctly and select the matching company name from the list.</p>");
```
Comment (richtier) [inline enrolment/static/javascripts/index.js]: putting HTML in javascript hast the disadvantages that it adds complexity when grepping for string (I can see a developer in  future going "oh, that html is added dynamically by index.js!" after scratching their head for a while)

to avoid that, we can move this html into the django html template, and hide is by default, and have this js instead selectively show/hide that error.

This also has the advantage that human-readable text is managed soley by django templates, meaning localisation and translations only need to happen in one place (the templates) rather than two (templates and js).

File: enrolment/static/javascripts/index.js [hunk 2]
```diff
@@ -631,16 +657,21 @@ GOVUK.page = (new function() {
   /* Find and apply a scroll in effect to specified element.
    **/
   function setupHomeScreenshotEffect() {
-    new GOVUK.effects.SlideIntoView($("#fabhome-screenshot"), 550);
+    var $homeScreenshot = $("#fabhome-screenshot");
+    if($homeScreenshot.length) {
```
Comment (richtier) [inline enrolment/static/javascripts/index.js]: I see this is checking that the element is on the page before running the code. The motivation for this appears to be a workaround of the fact that all the site's js is in one file.

A better solution is to split the js into multiple files.

File: enrolment/templates/landing-page.html [hunk 3]
```diff
@@ -19,9 +19,9 @@ <h1 class="element-within company-name-title">
 			<span>{% trans "Be promoted internationally" %}</span>
 			<span class="sub">{% trans "with an international trade profile" %}</span>
 		</h1>
-		<div {% if user_has_company %}style="visibility: hidden;"{% endif %}>
-			{% include 'register-company-number-form.html' %}
-		</div>
+
+		{% include 'register-company-number-form.html' %}
```
Comment (richtier) [inline enrolment/templates/landing-page.html]: we want to not show the form is the user already has a company.

File: enrolment/static/javascripts/index.js [hunk 4]
```diff
@@ -232,8 +233,14 @@ GOVUK.components = (new function() {
       $input.on("focus.SelectiveLookup", function() { instance._private.active = true; });
       $input.on("blur.SelectiveLookup", function() { instance._private.active = false; });
       $input.on("input.SelectiveLookup", function() {
+        if(instance._private.timer) {
```
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: The problem is, like Python, JS doesn't actually support public and private. We can mimic it with closures, but it is recommended to avoid those when using classes because it is more performant to place on the class the prototype. Generally, that avoids redeclarations of functions when creating instances. Using the prototype can also have the (objective) advantage of making it easier to read code, by not creating massive function (class) scopes. 

Since we then don't have full private facility using the recommended prototype declarations, we have to expose those variables for sharing across class methods (stupid, I agree, but unfortunately this is JS, not Java). 

We can follow the same convention as Python, which just adds an underscore to every 'should be' private variable, but the disadvantage of that is to pollute the namespace. 

The simple solution is to group those private variables under one namespace, e.g. _private. Doesn't have to be called that, but it does reenforce the intention for anyone not familiar with the underscore implied message. 

Basically, we're doing this:

instance
            bindContentEvents
            param
            setContent
            _private 
                         $errors
                         $field
                         $form
                         $input
                         $list
                         active
                         service
                         timer

Instead of this:

instance
            _$errors
            _$field
            _$form
            _$input
            _$list
            active
            bindContentEvents
            param
            setContent
            service
            timer

Which is difficult to replicate here, but usually makes things less cluttered and more readable when using a browser debugger or command line interrogation.

File: enrolment/static/javascripts/index.js [hunk 5]
```diff
@@ -232,8 +233,14 @@ GOVUK.components = (new function() {
...
+          clearTimeout(instance._private.timer);
+        }
+        
         if(this.value.length >= opts.lookupOnCharacter) {
-          instance.search();
+          instance._private.timer = setTimeout(function() {
```
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: Have you tried running that in IE8?

File: enrolment/static/javascripts/index.js [hunk 6]
```diff
@@ -416,20 +423,39 @@ GOVUK.components = (new function() {
    **/
   this.CompaniesHouseNameLookup = CompaniesHouseNameLookup;
   function CompaniesHouseNameLookup($input, $field) {
+    var instance = this;
     SelectiveLookup.call(this,
```
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: That's javascript inheritance. 

The SelectiveLookup is what you'd expect the functionality to be.
e.g. type in input and input receives the value of what you selected from dropdown 
- like you normally see in a predictive search input

CompaniesHouseNameLookup is extending that functionality as a custom handling variant.
e.g. type in input and a different (hidden in this case) field receives the value of what you selected from dropdown 

Doing this way means the core (more comment) functionality of SelectiveLookup should be reusable. 
Something I'd expect could be highly likely with MVPx version of the FAS search.

File: enrolment/static/javascripts/index.js (hunk 1 above)
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: I would not argue with the ideal preference of keeping HTML out of Javascript code, but this is a simple and common jQuery approach. 

The javascript (non-jQuery) approach looks nicer
e.g. 
```
var p = document.createElement("p");
var text = document.createTextNode("some message here");
p.appendChild(text);
instance._private.$errors.append(p);
```

but it's more long-winded and not really much different. 

Adding separate HTML inside the template is not JS enhancement. It explicitly attaches something that is JS component specific to the template. Lose the JS and later developers don't know why the HTML is there, as one negative impact example. 

I would not argue with having wording kept out of javascript, either. However, this site doesn't currently have an efficient mechanism for making strings available to the JS code, so I didn't implement one. Really, it's not that difficult but, along with the fact the known use of that code is for an English-speaking FAB home page, I chose not to go for the common/simple option. 
 
Incidentally, a simple way of keeping text strings out of JS, without peppering your HTML with lots of hidden messages to show/hide/ask jQuery to go find, is to have something like this in your <head>

```
<script>
GOVUK.messages = {
  chlookupError:  "{% trans "Please enter your company name" %}"
}
</script>
```

The JS would then look like this:
`
instance._private.$errors.append("<p>" + GOVUK.messages.chlookupError +"</p>");
`

File: enrolment/static/javascripts/index.js (hunk 2 above)
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: Exactly. 

The JS files should be split out. I started looking into it but stopped in favour of getting the required work done and worrying about refactoring later. 

It would be better to run different init() functions, perhaps from different page templates, or even by compiling JS into a single page and serving different versions, depending on page type. It's also common to put a page-specific HTML class name to reflect the page type, and kick off script init() functions based on what is detected. There are obviously different ways of handling the same thing. 

For this site, the current bug requirement, the sake of simplicity, and a quick solution, wrapping the specific code in a check to see if that required element was found, is easier than all the above.

File: enrolment/static/javascripts/index.js (hunk 1 above)
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: Should add, this error message should have aria-attributes to properly support accessibility. If you have time on your hands, feel free to add them or, if you prefer, raise a bug ticket and assign to me so it doesn't get lost.

File: enrolment/templates/landing-page.html (hunk 3 above)
Comment (stevenburnell-dit) [inline enrolment/templates/landing-page.html]: That's not in the Stories or designs. If you believe otherwise, please point me to the correct ticket and I will revisit.

File: enrolment/static/javascripts/index.js (hunk 1 above)
Comment (richtier) [inline enrolment/static/javascripts/index.js]: ~"lots of jQuery developers do it" eww :smile: 

 as you say - it would be an ideal preference of keeping HTML out of Javascript code. Please keep HTML out of Javascript. The following would be fine:

```
<div class="errors">
    {{ form.company_number.errors }}
    {# this error is shown/hidden according to javascript logic #}
    <ul class="errorlist js-validation-error-enter-company-name"><li>Please enter your company name</li></ul>
</div>
```

and then in the js code show/hide the class identified with ` js-validation-error-enter-company-name`.

File: enrolment/static/javascripts/index.js (hunk 6 above)
Comment (richtier) [inline enrolment/static/javascripts/index.js]: I think you're not getting my point.

In this situation:
```js
SelectiveLookup.call(this,
       $input,
        GOVUK.data.getCompanyByName
      );
```

results in the same outcome as:

```
SelectiveLookup(
        $input,
        GOVUK.data.getCompanyByName
);
```
no? If so, the simpler approach of not using `call` is preferable.

**Comment by richtier** [comment]
I suspect the issues wont be resolved today. we need this feature today. Merging as is..

File: enrolment/static/javascripts/index.js (hunk 4 above)
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: No, not quite. 

I could have used closures to prevent access to (intended private) variables outside the parent objects scope, such as the technique used with the Service class but, I decided to abide by the Mozilla recommended syntax of adding methods on the prototype for SelectiveLookup, for other known advantages.

It is true that the Mozilla recommended syntax is more performant because it avoids redeclaration of any function encapsulated within the function (class) scope, when creating instances, but I am unaware of any negative performance issue when comparing variables accessed from within a closure to those accessed by the object. 

My expectation, given that we are not creating many instances of the class, the performance difference would not be a deciding significant benefit on its own. 

Instead, another motivation to follow the Mozilla prototype usage was to try and limit the physical (line count) scope of the SelectiveLookup function and intentionally avoid creating too many, potentially confusing, closures. 

Unlike the Service class, which fits nicely onto one screen because it doesn't have a lot of code lines, the SelectiveLookup class is a lot bigger, with many more lines. It already spans more then one vertical page. 

In using the Mozilla recommend syntax I also gain the following code advantages:
1. SelectiveLookup function is reduced in number of lines.
2. Indentation (of nested methods) is reduced.
3. Context of _this_ should be easier to identify. 
4. Placing (required private) variables on the namespace makes it easier to understand where they are declared.
5. All of the above mean the need to scroll up and down is greatly reduced for other developers when trying to read and understand the code.

> and the namespacing hints to developers these variables are private
This is also true.

File: enrolment/static/javascripts/index.js (hunk 6 above)
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: No, it is not the same.

Using the second form will not result in the intended inheritance of SelectiveLookup since the value of _this_ inside the function will not be the same as the current _this_ (your crated instance).

The use of the Function.call() method is to explicitly set the _this_ value as the current _this_ value. 

In case I have again failed to explain in clearly, have a look at this simple example.

```
function A(str) {
  this._output = str;
  this.rawOutput = function () {
    console.log("raw: ", this._output);
  }
}
function B(str) {
  A.call(this, str);
  this.prettyOutput = function () {
    console.log('pretty: ' + this._output);
  }
}
function C(str) {
  A(this);
  this.formatOutput = function () {
    console.log('format: ' + this._output);
  }
}

var a = new A('Created from A');
var b = new B('Created from B');
var c = new C('Created from C');

a.rawOutput();     // Output "raw:  Created from A"
b.prettyOutput();  // Output "pretty: Created from B"
b.rawOutput();     // Output "raw:  Created from B"
c.formatOutput();  // Output "format: undefined"
c.rawOuput();      // Error "c.rawOuput is not a function"
```

Instance 'c' will not be assigned a `c._output property`, which is why you will get the 'undefined' output, and it will also not inherit the `rawOutput` method, resulting in the error.

File: enrolment/static/javascripts/index.js (hunk 1 above)
Comment (stevenburnell-dit) [inline enrolment/static/javascripts/index.js]: It would also be even more ideal to keep the frontend HTML out of the backend Python code, but we seem to have no problem with that. 

The Python output from variable...
{ form.company_number.errors }}

...is HTML written directly inside the python code. Now that is definitely wrong and plain confusing. 

Unless we're going to correct that, I don't think we need to worry about a working solution that keeps the frontend code within the frontend code, albeit in the less than ideal-world form, but still one very familiar and common to all jQuery/javascript/frontend developers.

To take and alter a quote from the original comment: 

> I can see a developer in future going "oh, that html is hardcoded in the utils.py!" after scratching their head for a while

Clearly it is not an ideal world but, at the cost of sticking with the simplistic, common practice of jQuery-style HTML additions, the code is adhering to Javascript enhancement, not splitting up functionality and dependencies. Basically, you only get the extra the html IF javascript is on AND the component has been created, AND the javascript functionality determines it should be displayed.

I think the only developers that might become even remotely confused over this, are backend developers that probably don't even need to worry themselves about it.
